### PR TITLE
Introduce RelationField / ScalarField split into the datamodel

### PR DESCRIPTION
**Pull request by do4gr**
Fixes https://github.com/prisma/prisma/issues/1612

### CODE REVIEW

File: libs/datamodel/core/src/dml/datamodel.rs [hunk 1]
```diff
@@ -94,16 +94,30 @@ impl Datamodel {
             .find(|m| m.fields().any(|f| f as *const Field == field as *const Field))
     }
 
+    /// Finds a model for a field reference by using reference comparison.
+    pub fn find_model_by_relation_field_ref(&self, field: &RelationField) -> Option<&Model> {
+        // This uses the memory location of field for equality.
+        self.models().find(|m| {
+            m.relation_fields()
+                .any(|f| f as *const RelationField == field as *const RelationField)
```
[the remark this answers is not in the document]
Comment (do4gr) [reply libs/datamodel/core/src/dml/datamodel.rs:102]: For relation fields I could circumvent this, for scalars I don't know how atm

[the remark this answers is not in the document]
Comment (do4gr) [reply libs/datamodel/core/src/dml/datamodel.rs:102]: The scalar one was unused anyways, so I'm removing it.

File: libs/datamodel/core/src/dml/field.rs [hunk 2]
```diff
@@ -152,29 +355,12 @@ impl WithDatabaseName for Field {
...
-            is_unique: false,
-            is_id: false,
-            documentation: None,
-            is_generated: false,
-            is_updated_at: false,
-            is_commented_out: false,
-        }
+impl WithDatabaseName for RelationField {
+    fn database_name(&self) -> Option<&str> {
+        None
```
[the remark this answers is not in the document]
Comment (do4gr) [reply libs/datamodel/core/src/dml/field.rs]: removed it.

File: libs/datamodel/core/src/dml/model.rs [hunk 3]
```diff
@@ -217,65 +255,33 @@ impl Model {
...
+        self.relation_fields().find(|rf| {
+            rf.relation_info.to == to
+                && rf.relation_info.name == relation_name
+                && (self.name != to || rf.name != exclude_field)
         })
     }
 
     /// Checks if this is a relation model. A relation model has exactly
     /// two relations, which are required.
     pub fn is_relation_model(&self) -> bool {
```
Comment (mavilein) [inline libs/datamodel/core/src/dml/model.rs:265]: This seems to be unused.

Comment (do4gr) [reply libs/datamodel/core/src/dml/model.rs:265]: Jep, removing it.

**Pull request merged**